numpy/ma/tests/test_core.py

```diff
@@ -1327,14 +1327,43 @@ def test_eq_on_structured(self):
         test = (a == a)
         assert_equal(test, [True, True])
         assert_equal(test.mask, [False, False])
+        test = (a == a[0])
+        assert_equal(test, [True, False])
+        assert_equal(test.mask, [False, False])
         b = array([(1, 1), (2, 2)], mask=[(1, 0), (0, 0)], dtype=ndtype)
         test = (a == b)
         assert_equal(test, [False, True])
         assert_equal(test.mask, [True, False])
+        test = (a[0] == b)
+        assert_equal(test, [False, False])
+        assert_equal(test.mask, [True, False])
         b = array([(1, 1), (2, 2)], mask=[(0, 1), (1, 0)], dtype=ndtype)
         test = (a == b)
         assert_equal(test, [True, False])
         assert_equal(test.mask, [False, False])
+        ndtype = [('A', int), ('B', [('BA', int), ('BB', int)])]
+        a = array([(1, (1, 1)), (2, (2, 2))],
+                  mask=[(0, (1, 0)), (0, (0, 1))], dtype=ndtype)
+        # complicated dtype, 2-dimensional array.
+        ndtype = [('A', int), ('B', [('BA', int), ('BB', int)])]
+        a = array([[(1, (1, 1)), (2, (2, 2))],
+                   [(3, (3, 3)), (4, (4, 4))]],
+                  mask=[[(0, (1, 0)), (0, (0, 1))],
+                        [(1, (0, 0)), (1, (1, 1))]], dtype=ndtype)
+        test = (a[0, 0] == a)
+        assert_equal(test, [[True, False], [False, False]])
+        assert_equal(test.mask, [[False, False], [False, True]])
+        # ensure simple examples are symmetric and make sense.
+        # from https://github.com/numpy/numpy/pull/8590#discussion_r101126465
+        m = mvoid((1, 2.), mask=(0, 1), dtype='i4,f4')
+        n = mvoid((1, 2.), dtype=m.dtype)
+        o = mvoid((1, 1.), mask=(0, 1), dtype=m.dtype)
+        assert_equal(m == n, False)
+        assert_equal(m == o, True)
+        assert_equal(n == o, False)
+        assert_equal(n == m, False)
+        assert_equal(o == m, True)
+        assert_equal(o == n, False)
```

Comment:
My argument would be that there are 3 cases here:

 * No matter what values the masked values take, two objects are always equal
 * No matter what values the masked values take, two objects are always not equal
* The equality depends on the values of the masked values

Since the first two of these cases are represented by `True` and `False`, surely we should use `masked` to represent the third case?